R/combine_checkboxes.R

```diff
@@ -40,87 +41,84 @@
                                tbl,
                                cols,
-                               values_to,
+                               names_prefix = "",
+                               names_suffix = NULL,
+                               names_sep = "_",
                                multi_value_label = "Multiple",
                                values_fill = NA,
                                raw_or_label = "label",
                                keep = TRUE) {
   # Check args ---
   check_arg_is_supertbl(supertbl, req_cols = c("redcap_data", "redcap_metadata"))
+  check_arg_is_character(names_prefix, len = 1)
+  check_arg_is_character(names_suffix, len = 1, null.ok = TRUE)
+  check_arg_is_character(names_sep, len = 1, any.missing = TRUE)
   check_arg_is_character(tbl, len = 1, any.missing = FALSE)
-  check_arg_is_character(values_to, len = 1, any.missing = FALSE)
   check_arg_is_character(multi_value_label, len = 1, any.missing = TRUE)
   check_arg_is_character(values_fill, len = 1, any.missing = TRUE)
   check_arg_choices(raw_or_label, choices = c("label", "raw"))
   check_arg_is_logical(keep, len = 1, any.missing = FALSE)
 
-  # Save user cols to enquosure
-  cols_exp <- enquo(cols)
-
   # Extract tbl from supertbl
   data_tbl <- supertbl %>%
     extract_tibble(tbl)
 
-  # Get field names from cols_exp, check that fields exist
-  field_names <- names(eval_select(cols_exp, data = data_tbl))
-  check_fields_exist(fields = field_names, expr = cols_exp)
+  # Save user cols to quo
+  cols_exp <- enquo(cols)
 
-  # Define values_to as the count of TRUEs/1s for the given checkbox field
-  # Assign TRUE if multiple selections made, and FALSE if one or zero made
-  data_tbl_mod <- data_tbl %>%
-    mutate(
-      !!values_to := case_when(
-        rowSums(select(., eval_tidy(cols_exp))) > 1 ~ TRUE,
-        TRUE ~ FALSE
-      )
-    )
+  # Evaluate the cols expression to get the selected column names
+  selected_cols <- names(eval_select(cols_exp, data = data_tbl))
+  check_fields_exist(fields = selected_cols, expr = cols_exp) # Check supplied fields exist
 
   # Get metadata reference table, check that chosen fields are checkboxes
-  metadata <- get_metadata_ref(data_tbl_mod, supertbl, tbl, field_names)
+  metadata_tbl <- supertbl$redcap_metadata[supertbl$redcap_form_name == tbl][[1]]
+  metadata_spec <- get_metadata_spec(metadata_tbl, selected_cols, names_prefix, names_suffix, names_sep)
 
-  # Replace TRUEs/1s with raw/label values from metadata
-  data_tbl_mod <- data_tbl_mod %>%
-    mutate(across(
-      field_names,
-      ~ replace_true(.x,
-        cur_column(),
-        metadata = metadata,
-        raw_or_label = raw_or_label
+  # Define .new_col as the count of TRUEs/1s for the given checkbox field
+  # Assign TRUE if multiple selections made, and FALSE if one or zero made
+  data_tbl_mod <- data_tbl
+  .new_col <- unique(metadata_spec$.new_value)
+
+  for (i in seq_along(.new_col)) {
```

Comment:
Consider abstracting the logic from here to L116 into a single function that does all the transformations for a single output column. If I was writing it I would probably have it take:
- A tibble containing `field_name`, `raw`, `label`
- A string with a single `.new_value`
- The data tibble
and return a one column `tibble` with the new column

Then you could do:

```r
combine_checkboxes_impl <- function(metadata, .new_value, data_tbl) {
...
}

new_cols <- metadata_spec |>
  nest(.by = .new_value, .key = "metadata") |>
  pmap(combine_checkboxes_impl, data_tbl = data_tbl)

bind_cols(data_tbl, new_cols)
```

Reply:
Great idea, will work on implementing.

Reply:
Recent commit should address this! Consolidated a lot of code in the end 👍